### DISCUSSION TITLE
On-Demand Emails to Volunteers

### DISCUSSION THREAD

**Discussion by ArturoTrailblazer** [Sundae-Shop-Consulting/Volunteers#37] [Feature Design]
What are the high-level feature pros and cons for the two email options, BBC Send Email Action v Custom LWC

**Comment by dospromptman** [comment]
Arturo, thank you for starting this discussion!

For anyone new coming to this discussion, I'll clarify the scope: we're discussing how it may be possible to achieve "ad-hoc email" capabilities, i.e., allowing a volunteer manager to visit a volunteering record (a Volunteer Position, Volunteer Shift, or Volunteer Activity) and click a "Send Email" button which then lets them send an email to each volunteer currently assigned to that volunteer record.

The scenarios where this might be used might be: to send an email to all volunteers assigned to a specific position letting them know about an interesting volunteer shift or activity that is coming up; to send an email to all volunteers associated with a specific volunteer shift or activity letting them know last minute details about where to go when they arrive, or a reminder to bring certain gear (i.e. "It is raining today, don't forget your umbrella"), etc.

The functionality in question should allow the volunteer manager to focus on composing the email to the volunteers instead of figuring out who all is assigned and manually addressing those individuals in any other way.

The options being considered for supporting this use case are:  **Delegate to Existing Send Email Global Action** and **Custom Email Composer**.

## Delegate to Existing Send Email Global Action

This approach involves writing a small amount of custom code in the form of a custom quick action for the record that should support emailing as well as a custom Lightning Web Component which understands how to query for the email addresses of all volunteers associated with that record.  When the volunteer manager invokes the action on the volunteer record, the custom LWC then uses the list of emails to invoke the existing Send Email global action which Salesforce provides which opens an email composer allowing the user to author and send a single email message.

### Pros

- Simpler to implement (less code, less moving pieces, less responsibility for critical functionality) than fully custom solution-- although this solution still requires writing some custom code to support
- Users get all the functionality available using the standard Send Email global action, including:
  - selection of sender address as current user or from allowed org wide email addresses,
  - the full featured email composer rich text editor (including inserting images),
  - adding attachments,
  - the merge syntax picker,
  - loading templates and creating new or saving templates,
  - quick text functionality,
  - previewing the email,
  - and any new enhancements which Salesforce might add to this composer (i.e. AI assisted authoring?).
- Associates sent email with volunteer record as an activity
- Works better with "[Send Email from Salesforce with Gmail™ or Office 365™](https://help.salesforce.com/s/articleView?id=sf.emailadmin_sendthroughexternal_overview.htm&type=5)"-- this is relevant because it might be easier for small orgs to set up.
- Theoretically more relaxed on email sending limits-- supposedly emails sent using the UI do not count against org sending limits, but I haven't tested this to ensure this is true so this needs to be validated before considering for discussion.

### Cons

- Because the Send Email global action is not a mass email editor, by its nature it only allows sending of a single email.  This means the email should be sent to a single address (probably the sender's email address) and all additional recipients need to be added as 'bcc' recipients.
- Limited number of recipients are allowed in this composer (haven't tested this exactly, but some documentation suggests the limit could be around 100 recipients?).  This needs to be tested before considering for discussion.
- Send Email global action only allows providing recipients as a list of email addresses, so the emails sent are not necessarily explicitly associated with Contact (or other) records-- because the email is only sent to one actual recipient, this limits the ability to track the email as an activity for each volunteer recipient (including open tracking, bounce tracking, etc).  Recipients for sent emails can still be viewed on the related email activity detail page, but recipients in bcc will show email only.
- Because the email is technically sent to only one recipient, use of Recipient merge syntax is not useful (and in some cases will break in the standard composer if an email address is used for the To recipient rather than a Salesforce record (like a Contact or User).  This means emails sent this way can not be customized for the recipient in any way.  Merge fields to include information for the related volunteer record can be included, though.

## Custom Email Composer

This approach involves writing custom quick actions, custom Apex controllers, and custom LWCs to fully implement an email composer UI, backend email send logic, and logic which can understand how to identify and treat as email recipients all volunteers associated with any of the supported objects (Volunteer Position, Shift, Activity).  When the volunteer manager invokes the action on the volunteer record, a custom UI is shown which pre-populates the recipients as a list of Contacts and allows them to author and send a single email message which is individually sent to each listed recipient.

### Pros

- Full control over the email sending experience, both in initial release and allowing flexible evolution in the future (contrasting to the delegation approach, where we could never provide functionality different than what Salesforce offers for sending single emails).
- Emails are sent individually to volunteer Contact records, so activities are tracked appropriately for each email sent to each recipient.  This means you also get bounce management, mail open tracking, etc, on a per recipient basis.
- Emails sent are associated with volunteer record as activities.
- Theoretically should be able to send to larger numbers of recipients than delegation approach (but scalability testing has not yet been done on the exact upper limit).
- A subset of email composer features are expected to be available in initial release:
  - selection of sender address as current user or from allowed org wide email addresses,
  - most functionality of the rich text editor is available, plus color and font selection, but minus the ability to include images in emails
- supports merge syntax for both the volunteer record and the recipient in the message body, so emails can be customized for the recipient

### Cons

Since this approach includes the ability to evolve the feature, the following represents features missing from an expected MVP release, but in many cases can be mitigated with future development/enhancement.

- Can't include images in rich text editor
- Can't add attachments to the email
- Can't use the quick text feature
- Can't choose from templates or create new or update existing templates
- No preview email functionality
- Can't be used with "Send with Gmail or Outlook" functionality-- emails sent from Apex are always sent by Salesforce's mail servers.  This has implications in set up, because users must ensure that the email address they use as the sender is properly configured to allow DMARC validation to pass when Salesforce sends the email-- this is not possible if the user is using a standard gmail email address as the email address on their user or in an org wide email address (but should work if they are using Google Workspaces or a custom domain email address, if it is properly configured).  Note this same configuration burden applies to the standard Send Email global action, in the case where you are NOT using "Send with Gmail or Outlook" functionality, too.
- Emails sent using custom functionality are subject to daily org email limits.

**Comment by dospromptman** [comment]
In our conversation earlier, @ArturoTrailblazer, it sounds like you might consider the ability to choose an existing email template and possibly also the ability to insert merge fields via UI chooser to be critical pieces missing from the custom approach and worth considering as a push to include in MVP?

I think @lmeerkatz mentioned that if the ability to choose an existing template existed, then the user can always use the standard email template composer to insert merge fields as well, so is the main thing missing then just the ability to choose an existing template?

**Comment by ArturoTrailblazer** [comment]
So I thought I posted a message about [DataFetcher](https://appexchange.salesforce.com/appxListingDetail?listingId=b7953014-b109-4ab6-a721-37e9969e67f5) and Skye Tyler uses this for several projects and I was connecting with her on a project we worked on and she mentioned that it could work for supporting some of this need as well - she's also used it for some matching use cases that may be in future build needs for this.  @dospromptman I'll look over your comments as well, but still learning the Github tool and so hadn't seen them before now.  

If you're not familiar with Data Fetcher, Skye is at MidWest Dreamin this week, but I could probably connect with her next week to see if she could do a walk through of her demo that she's built out using the solution.

**Reply by lmeerkatz** [reply]
Thanks, @ArturoTrailblazer! I just watched the demo and it looks like DataFetcher is just providing an easier way to query records in a flow. Is there something else that it does that you were thinking would be a good fit for emailing volunteers?

**Reply by ArturoTrailblazer** [reply]
no @lmeerkatz Just wondering if it made for an easier approach to address some of the concerns when used in conjuction with merge fields, customizing text/templates, etc. to the two approaches.

**Comment by ArturoTrailblazer** [comment]
@dospromptman One thing that I've run into with two clients lately when using Send List Email is if the contact is marked as Email Opted Out they are included in the number of recipients but excluded from the actual send.  Both these clients use Pardot/MCAE and are syncing the Pardot Unsubscribe/Opt Out to the Salesforce Email Opt Out field and that is what is causing the issue for them and wanted to bring this to your attention in case it could impact these solutions/approaches.  

We are able to enable the Send Non-marketing/Operational Emails permission to a user but that will only work when sending a Salesforce Email from the contact's record and does not extend to the Send List Email feature (this is verified by the SF Support team).  I know it isn't something that we can control per say, but want to make sure we call these things out in any documentation that we provide that if the Contact is Email Opt Out they may not receive this on demand email and that folks should consider how they are managing their opt outs or if we should call out "hey these people that you selected have opted out of emails" and therefore can't be included and you must reach out to them individually or through other channels.  cc: @lmeerkatz
